### PR TITLE
C#: Simplify dispatch hierarchy for `getLocation()`

### PR DESCRIPTION
**Pull request by hvitved**

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/Location.qll [hunk 1]
```diff
@@ -171,6 +171,9 @@ class Assembly extends Location, Attributable, @assembly {
   /** Gets the version of this assembly. */
   Version getVersion() { assemblies(this, _, _, _, result) }
 
+  /** Gets an attribute attached to this assembly, if any. */
+  final Attribute getAnAttribute() { attributes(result, _, this) }
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/Location.qll]: `result = this.(Attributable).getAnAttribute()` ?

Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/Location.qll]: Doesn't work, as `Assembly` is no longer a sub type of `Attributable`.

**Comment by hvitved** [comment]
> What is the performance of this?

I'll run a dist comparison and post the results here when they are ready.

File: csharp/ql/src/semmle/code/csharp/Location.qll (hunk 1 above)
Comment (calumgrant) [reply csharp/ql/src/semmle/code/csharp/Location.qll]: Won't it cast via the db type?

Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/Location.qll]: Nope, it will cast via `Attributable` which will restrict it to `Element`.

**Comment by hvitved** [comment]
Much to my surprise, this seems to cause a slowdown on some projects ([internal link](https://git.semmle.com/gist/tom/61eeabca960c98beb5f29c8f14ffbedb)), so I am parking this PR for now.

**Pull request closed without merging**